### PR TITLE
Function Inlining

### PR DESCRIPTION
**Pull request by aviralg**
An implementation of function inlining which inlines functions in the program until it encounters recursive calls. OSR is not handled correctly.

### CODE REVIEW

File: parser.messages
Comment (gasche) [inline parser.messages]: The changes look like something your editor would do automatically and they are wrong (they turn unbreakable spaces into breakable spaces). Could you revert the changes to `parser.messages`? (`git show master:parser.messages > parser.messages` would revert the file, and then you could commit that.)

Comment (aviralg) [reply parser.messages]: Done.

File: analysis.ml
Comment (o-) [inline analysis.ml:183]: I know this is not really you code. But this looks a bit fishy. I think we should:
1. Check for duplicate parameters in the `check.ml` module and not here
2. only have one of the two `as_var_list` or `as_var_set` (I don't care which one) and then when you need something different the caller just needs to `VarSet.of_list` or `VarSet.elements`

File: edit.ml
Comment (o-) [inline edit.ml]: You can do `| Label (MergeLabel l | BranchLabel l) -> ...`

File: instr.ml
Comment (o-) [inline instr.ml:1]: haha, thanks :)

File: instr.ml
Comment (o-) [inline instr.ml]: maybe `replace_fun`

File: transform_inline.ml
Comment (o-) [inline transform_inline.ml:5]: I think we should probably change the type of analysis input. Obviously a set is not always optimal.

File: transform_inline.ml
Comment (o-) [inline transform_inline.ml]: It seems @gasche 's `Instr.map` should be used here. See `Edit.replace_uses` for an example on how to use it. The idea is that you inherit from `Instr.map` and override all the functions that correspond to the parts of instructions you want to change.

File: transform_inline.ml
Comment (o-) [inline transform_inline.ml:82]: I guess at this point the scope of the callee is guaranteed to be still correct in itself, because you did just renamings, right? But we got to be extra careful since it will fail on an ill-scoped function.

File: transform_inline.ml
Comment (o-) [inline transform_inline.ml]: maybe you could emmit an `Assert (false)` instruction as well?

**Review by o- — COMMENTED** [review]
sweet!

lgtm, except for the comments I put inline.

Don't forget to `git rebase --root -i` and squash all commits into (probably) one.

File: examples/inline_advanced.sou
Comment (mhyee) [inline examples/inline_advanced.sou]: For these example programs, you can use `assert`s instead of `print`s.

File: examples/inline_advanced.sou
Comment (mhyee) [inline examples/inline_advanced.sou:1]: Are the inlined versions really big? If not, it would be nice to include them in the examples directory, so we can see the before/after programs.

File: examples/inline_advanced.sou
Comment (mhyee) [reply examples/inline_advanced.sou]: I see that you're using `read_return` in the test suite. But using `assert` here allows us to test the programs with the test runner, which tries all combinations of optimizations.

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: Nitpick: please put space before and after `:`.

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml:24]: Should this be in `instr.ml` where `all_defined_vars` and `all_dropped_vars` are defined?

Actually, are `all_defined_vars` or `all_dropped_vars` even used?

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: This is a style nitpick, but we've been generally using `head :: tail` instead of `List.cons head tail`.

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: It's more verbose, but I think it'd be clearer if you wrote (with the shorter case first)

```ocaml
let replace x =
  if not (List.mem x vars_alist)
  then x
  else List.assoc x vars_alist
```

Appending `(x,x)` feels too clever for me.

File: transform_inline.ml
Comment (mhyee) [reply transform_inline.ml]: Yeah, looking over the function, I think all you need to override are `variable_use`, `variable_assign`, and `binder`.

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: Nitpick: missing space before `=`.

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: Please elaborate on what "information" is.

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: `fresh` instead of `Edit.fresh`.

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: It makes me a bit nervous that `fresh` is being called a second time here, and that the renamed formals will match up with the renamed function variables. If `fresh` was implemented as something like `gensym` that kept a global counter, then this would break.

`replacements` should include renamings for the formals, so maybe you could use that? Again, it would be more verbose, but then you wouldn't need to call `fresh` twice.

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: Nitpick: missing space before `=`.

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: Mention `ret_lab` in the comment?

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: You could probably use `Instr.map` here and override `goto_label` and `branch_label`.

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: I think it's fine to shadow, i.e. use `let callee = callee |> replace_vars caller ...` since you never use the original `callee` again.

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: Couldn't you combine the callee and caller instructions, so that `fresh_label` is only called once?

Similarly for `res_var` below.

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: Nitpick: missing space before `=`.

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: Can you think of better names? It's a bit messy to have, e.g. `cle`, `callee`, `callee'`, and `callee''`. (Personally, I find even `clr` and `cle` to be too terse.)

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: I'd prefer `caller` and `callee`. `clr` and `cle` don't save all that much.

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: `(clr.name, f, dep) :: ord`

File: transform_inline.ml
Comment (mhyee) [inline transform_inline.ml]: I'm having a hard time following this function. (It doesn't help that it's late and this is the last function I have to review.)

In particular, I can't follow the terse names like `ord`, `clr`, `vis`, `red`, and `dep`. (`acc` is commonly used so it's fine.)

Also, passing a 5-tuple around is kind of awkward; I'd be tempted to just define a new record type. That would help with readability, and also in the places where you only need to change one field.

Finally, I think it's a bit excessive when you have names like `ord`, `ord'`, `ord''`, and `ord'''`.

**Review by mhyee — CHANGES_REQUESTED** [review]
This is really cool! And thanks for the detailed comments!

I'd like to see an example where constant folding (or some other optimization) can be applied after inlining. Right now, none of the inlining examples allow for multiple optimizations.

I'd like to try running inlining on the vector example, but it needs to be rebased/updated, so I think I'll wait until this PR is merged.

I'm requesting changes because I'd like to look at `compute_inlining_order` again, when I'm less tired and its names are better.

File: analysis.ml
Comment (aviralg) [reply analysis.ml:183]: `as_var_list` has the advantage of maintaining the order of `formals` which is lost as soon as conversion to `Set` is performed. If we intend to change the type of `analysis_input` to contain a `List` of `formals` rather than a `Set` of `formals`, then `as_var_list` makes more sense.

File: instr.ml
Comment (aviralg) [reply instr.ml]: I did that because there is a function called `lookup_fun` and `update` seems to go along well with `lookup`. But its really a minor thing. Changed it to `replace_fun`.

File: transform_inline.ml
Comment (aviralg) [reply transform_inline.ml]: Good catch, I intended to replace that, but forgot. Corrected it now.

File: transform_inline.ml
Comment (aviralg) [reply transform_inline.ml]: This one looks a bit long and requires 2 lookups. A better version would be the following -
```
try List.assoc x vars_alist with
| _ -> x
```
I had this version initially but felt that the clever approach was readable.

File: examples/inline_advanced.sou
Comment (aviralg) [reply examples/inline_advanced.sou:1]: They are big.

File: instr.ml
Comment (aviralg) [reply instr.ml:1]: Welcome :-) . Gabriel suggested this while I was unifying the freshen family of functions. He said that - "its dirty but its you who is doing it". I guess the joke's on him now ;-)

File: transform_inline.ml
Comment (aviralg) [reply transform_inline.ml:5]: I can do that in a separate PR.

File: transform_inline.ml
Comment (aviralg) [reply transform_inline.ml:24]: Removed them. They are not used.

Comment (aviralg) [reply transform_inline.ml:24]: This will be moved to `instr.ml` once we change `analysis_input` to contain a `List` of `formals`.

File: transform_inline.ml
Comment (aviralg) [reply transform_inline.ml:82]: Yup. That's why `replace_vars` also replaces `formals`' `names`.

**Review by aviralg — COMMENTED** [review]
Thanks for the reviews guys. I believe I have resolved most of the issues.

File: transform_inline.ml
Comment (aviralg) [reply transform_inline.ml]: The idea was to use `clr` and `cle` for just the identifiers and not the actual objects.

File: transform_inline.ml
Comment (aviralg) [reply transform_inline.ml]: Done.

**Comment by gasche** [comment]
> A better version would be the following -
>
>     try List.assoc x vars_alist with
>     | _ -> x

Catching all exceptions is not very good style, it's better to be specific in the exception name

    try List.assoc x vars_alit with Not_found -> x

**Comment by gasche** [comment]
In the interest of keeping the main repository somewhat bisectable, I would be interesting in having the series of commits reworked into a few well-separated commits such that each commit leaves the repository in a correct (buildable, testable) state. If you wish, we can do this together on Monday morning. I feel that the PR is in an essentially mergeable state now, but I like this to happen before merging.

**Comment by mhyee** [comment]
Thanks for the updates. I still have a few remaining comments:

1. Is the `inline_basic` example small enough that you could add the before/after?

2. I'd like to see an inlining example that allows constant folding (or some other optimization).

3. The extra documentation for `compute_inlining_order` helps, but I'm still going to push for using a record type and using better variable names. (As you've probably noticed, I'm not a big fan of terse names, and I think a name like `foobar''` (with two primes) is already going too far.)

> In the interest of keeping the main repository somewhat bisectable, I would be interesting in having the series of commits reworked into a few well-separated commits such that each commit leaves the repository in a correct (buildable, testable) state. If you wish, we can do this together on Monday morning. I feel that the PR is in an essentially mergeable state now, but I like this to happen before merging.

Would squashing everything into a single commit be OK, or do you prefer more granular commits? I'd be fine with a single commit -- it would be a large commit, but I don't see any obvious place to break it up.

**Comment by gasche** [comment]
I'm ok with a big squash, but we could also split (eg. the `fresh` refactoring can go separately).

File: analysis.ml
Comment (o-) [reply analysis.ml:183]: agreed

**Comment by o-** [comment]
Squash, merge, fix the rest later, and repeat :)

**Comment by aviralg** [comment]
@mhyee - Regarding the `compute_inlining_order` function -
1) The initial version had full names rather than mnemonics, but that was causing quite a few splits in the lines to respect the 80 character limit. I chose these mnemonics because they are obvious and make the code compact enough without exceeding the screen size.
2) I don't see a compelling reason to use a record here. This is a perfect case for a `tuple`. I don't want to create a record type just for one function. Besides it would be used no differently than a `tuple`. 
3) The terse names are mnemonics and the documentation gives the full forms. `vis` for visited and `red` for reduced are readable.
4) I renamed some variables, I believe most of your complaints will be gone now.

**Comment by aviralg** [comment]
@gasche  @mhyee  @o-  I would squash all in a single commit just because its quick, easy and also because the change to `fresh` family of functions is minor refactoring. But I am also open to having 2 commits, one that refactors `fresh` and friends, and the other that implements inlining, though that might be more difficult for me due to my poor `git-fu` (I would be happy to know the ways of the masters).

**Comment by mhyee** [comment]
I still disagree about `compute_inlining_order`, but in the interest of moving forward, I'll approve the changes.

**Review by mhyee — APPROVED** [review]

**Comment by gasche** [comment]
👏 

(Rebase didn't work, we had to `export HASH=$(git rev-parse HEAD); git reset master; git commit -a -C $HASH`.)

**Comment by o-** [comment]
ah probably because of the merge commits in there...

**Pull request merged**
